cluster-autoscaler/cloudprovider/juju/juju_manager.go

```diff
@@ -122,27 +134,40 @@ func (m *Manager) refresh() error {
 	// This could mean updating the state of units currently managed by the manager
 	// or incorporating a totally new unit that was added by the cluster-admin manually
 	for unitName, unitStatus := range fullStatus.Applications[m.application].Units {
-		// Check if we are already managing this unit
-		if _, ok := m.units[unitName]; ok {
-			// Update the status and hostname (if it was empty) of each unit
-			m.units[unitName].status = unitStatus
-			if m.units[unitName].kubeName == "" {
-				m.units[unitName].kubeName = fullStatus.Machines[unitStatus.Machine].Hostname
+
+		unitHostname := fullStatus.Machines[unitStatus.Machine].Hostname
+		unitProviderID := ""
+		if unitHostname != "" {
+			node, err := m.kubeClient.CoreV1().Nodes().Get(ctx.TODO(), unitHostname, v1.GetOptions{})
+			if err != nil {
+				klog.Errorf("error getting provider ID for unit %v with hostname %v: %v", unitName, unitHostname, err.Error())
+			} else {
+				unitProviderID = node.Spec.ProviderID
 			}
-		} else {
+		}
+
+		// Check if we aren't already managing this unit
+		if _, ok := m.units[unitName]; !ok {
 			// Check if the unit is active and idle
 			// This is necessary since when a unit gets deleted it does not happen immediately
 			// We want to make sure we only add externally added units, not recently deleted units that are still showing up in status
 			if unitStatus.WorkloadStatus.Status == "active" && unitStatus.AgentStatus.Status == "idle" {
 				// The unit was added manually. Need to add it to the units list as a new unit
 				m.units[unitName] = &Unit{
-					state:    cloudprovider.InstanceCreating,
-					jujuName: unitName,
-					status:   unitStatus,
+					state:      cloudprovider.InstanceRunning,
+					jujuName:   unitName,
+					hostname:   unitHostname,
+					status:     unitStatus,
+					providerID: unitProviderID,
```

Comment:
is it ok that this could still be `""` if the lookup above failed?
what happens if `node.Spec.ProviderID == nil`?

Reply:
since the provider ID is a string, it cant be nil, it can only either have a value or be "". Its okay if providerID is "", that would mean its just not been added as a node in k8s yet and is still coming up. The autoscaler likes to know about nodes in all states, even if they arent technically a k8s node yet. Units with states of InstanceCreating would not have providerIDs set off the bat

Reply:
Thanks so much for the details.  LGTM